source/material_model/steinberger.cc

```diff
@@ -501,21 +500,41 @@ namespace aspect
       double cp = 0.0;
       if (!latent_heat)
         {
-          if (n_material_data == 1)
-            cp = material_lookup[0]->specific_heat(temperature,pressure);
+          if (material_file_names.size() == 1)
+            {
+              cp = material_lookup[0]->specific_heat(temperature,pressure);
+            }
+          else if (material_file_names.size() == compositional_fields.size() + 1)
+            {
+              const double background_cp = material_lookup[0]->specific_heat(temperature,pressure);
+              cp = background_cp;
+              for (unsigned int i = 0; i < compositional_fields.size(); ++i)
+                cp += compositional_fields[i] *
+                      (material_lookup[i+1]->specific_heat(temperature,pressure) - background_cp);
+            }
           else
             {
-              for (unsigned i = 0; i < n_material_data; i++)
+              for (unsigned i = 0; i < material_file_names.size(); i++)
                 cp += compositional_fields[i] * material_lookup[i]->specific_heat(temperature,pressure);
             }
         }
       else
         {
-          if (n_material_data == 1)
-            cp = material_lookup[0]->dHdT(temperature,pressure);
+          if (material_file_names.size() == 1)
```

Comment:
To me it looks like you won't need this special case but you can handle it in the ``else`` branch below (if you only do the sum if you have ``n_c+1`` files). This holds in all other places too.